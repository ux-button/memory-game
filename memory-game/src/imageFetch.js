const fetchNames = async (quantity) => {
    try {
        const responce = await fetch(`https://pokeapi.co/api/v2/pokemon?limit=${quantity}`, {mode: 'cors'});
        const pokemonData = await responce.json()
        const names = await pokemonData['results'].map(result => result['name']);
        return names
    } catch (error) {
        // Here error handler
    }
}

const fetchImageUrl = async (name) => {
    try {
        const respond = await fetch(`https://pokeapi.co/api/v2/pokemon/${name}`, {mode: 'cors'});
        const pokemonData = await respond.json();
        return pokemonData['sprites']['front_default']
    } catch (error) {
        // Here error handler
    }
}

const fetchExtra = async (extra, initial=0) => {
    const full = initial + extra;

    // Fetch names
    const pokemonNames = await fetchNames(full);

    // Fetch urls
    const imagesUrl = [];
    for (let name of pokemonNames) {
        const url = await fetchImageUrl(name);
        imagesUrl.push(url)
    }

    // Combine in single object
    const extraPokemons = [];
    for (let i = initial; i < full; i++) {
        extraPokemons.push({id: i, name: pokemonNames[i], url: imagesUrl[i]})
    }

    return extraPokemons
}

export { fetchExtra }