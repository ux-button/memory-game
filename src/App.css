body {
    margin: 0;
    padding: 0;
    background-color: #222222;
    font-family: system-ui, -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif;
}

.navigation-bar {
    display: flex;
    justify-content: center;
    align-items: center;
    height: 40px;
    background-color: black;
    color: white;
}

.card-board {
    padding: 16px;
    display: grid;
    grid-template-columns: repeat(auto-fill, minmax(300px, 1fr));
    gap: 16px;
}

.card {
    color: white;
    background-color: #333;
    border-radius: 16px;
    box-shadow: 1px 5px 17px -1px rgba(0,0,0,0.05);
    padding: 16px 0;
    display: flex;
    flex-direction: column;
    align-items: center;
    will-change: transform;
    transition: background-color 100ms ease-out, transform 100ms ease-out;
}

.card:hover {
    background-color: #444;
    transform: translateY(-4px);
    cursor: pointer;
}

@media (max-width: 500px) {
    .card-board {
        grid-template-columns: repeat(auto-fill, minmax(150px, 1fr));
    }
  }