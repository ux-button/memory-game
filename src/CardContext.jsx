import { createContext } from "react";

const simpleData = createContext({
    days: "Hey, I'm context",
    ballons: (day) => {
        return `${day * 5} ballons`
    }
})

export { simpleData }