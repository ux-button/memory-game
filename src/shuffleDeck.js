const shuffleDeck = (cards) => {
    cards.sort(() => Math.random() - 0.5);
}

export { shuffleDeck }
  
  