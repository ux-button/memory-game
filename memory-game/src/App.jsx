import { useState, useEffect, useRef, useInsertionEffect } from 'react'
import { fetchExtra } from './imageFetch'
import { shuffleDeck } from './shuffleDeck'
import { Card } from './components/Card'
import { NavigationBar } from './components/Navigation'
import './App.css'

export function Game () {
  const [pokemons, setPokemons] = useState([]);
  const [score, setScore] = useState({points: 0, best: 0});

  const match = useRef([]);
  let counter = useRef(0)

  useEffect(() => {
    async function getExtra (extra, initial) {
      const fetchExtraPokemons = await fetchExtra(extra, initial);
      shuffleDeck(fetchExtraPokemons)
      setPokemons(fetchExtraPokemons);
    }
    getExtra(30, pokemons.length)
  }, [counter.current])

  function checkMatch (id) {
    if (match.current.includes(id)) {
      // Game over state
      match.current = [];
      if (score.points > score.best) {
        setScore({...score, points: 0, best: score.points});
      } else {
        setScore({...score, points: 0});
      }
      shuffleDeck(pokemons);
    } else {
      // Score state
      match.current.push(id);
      setScore({...score, points: score.points + 1});
      shuffleDeck(pokemons);
    }
  }

  const cardsPile = pokemons.map(pokemon => {
      return (
        <Card key={pokemon.id} id={pokemon.id} pokemonName={pokemon.name} imageUrl={pokemon.url} handleClick={checkMatch}/>
      )
  })

  return (
    <>
      <div className='navigation-bar'>
        <NavigationBar score={score} />
      </div>
      <div className='card-board'>
        {cardsPile.slice(0, 15)}
      </div>
    </>
  )
}