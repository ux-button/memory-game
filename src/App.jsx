import { useState, useEffect, useRef, useInsertionEffect } from 'react'
import { fetchExtra } from './imageFetch'
import { shuffleDeck } from './shuffleDeck'
import { Card } from './components/Card'
import { NavigationBar } from './components/Navigation'
import trobber from './assets/trobber.svg'
import './App.css'

export function Game () {
  const [state, setState] = useState('loading');
  const [pokemons, setPokemons] = useState([]);
  const [score, setScore] = useState({points: 0, best: 0});

  const match = useRef([]);

  useEffect(() => {
    let ignore = false;

    async function getExtra (extra, initial) {
      const fetchExtraPokemons = await fetchExtra(extra, initial);
      if (!ignore) {
        shuffleDeck(fetchExtraPokemons);
        setPokemons(fetchExtraPokemons);
        setState('ready')
      }
    }
    getExtra(150, pokemons.length)

    return () => {
      ignore = true;
    };
  }, [])

  function checkMatch (id) {
    if (match.current.includes(id)) {
      // Game over state
      match.current = [];
      if (score.points > score.best) {
        setScore({...score, points: 0, best: score.points});
      } else {
        setScore({...score, points: 0});
      }
      shuffleDeck(pokemons);
    } else {
      // Score state
      match.current.push(id);
      setScore({...score, points: score.points + 1});
      shuffleDeck(pokemons);
    }
  }

  const cardsPile = pokemons.map(pokemon => {
      return (
        <Card key={pokemon.id}
          id={pokemon.id}
          pokemonName={pokemon.name}
          imageUrl={pokemon.url}
          handleClick={checkMatch}/>
      )
  })

  if (state === 'loading') {
    return (
      <>
        <div className='nav'>
            <NavigationBar score={score} />
        </div>
        <div className='loading-container'>
          <img src={trobber}></img>
          Loading...
        </div>
      </>
    )
  } else if (state === 'ready') {
    return (
      <>
        <div className='nav'>
          <NavigationBar score={score} />
        </div>
        <div className='card-board'>
          {cardsPile.slice(0, 12)}
        </div>
      </>
    )
  }
}