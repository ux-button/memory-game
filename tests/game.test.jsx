import { render, screen } from '@testing-library/react';

import { Game } from '../src/App';

describe('App', () => {
  it('renders headline', () => {
    render(<Game title="React" />);

    screen.debug();

    // check if App components renders headline
  });
});