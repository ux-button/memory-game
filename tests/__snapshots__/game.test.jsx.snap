// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Game component > render snapshot correctly 1`] = `
<div>
  <div
    class="nav"
  >
    <div
      class="navigation-bar"
      role="textbox"
    >
      Score: 
      0
       âœ¦ Best score: 
      0
    </div>
  </div>
  <div
    class="loading-container"
  >
    <img
      src="/src/assets/trobber.svg"
    />
    Loading...
  </div>
</div>
`;
